feat - incorporated mcp server request configs for cli mode

Kavyapriya-1804 opened the pull request:
<!-- Provide a brief summary of your changes -->
This PR implements the feature requested by #330 

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->
When using the Inspector in `CLI mode`, there's currently no way to configure Inspector-specific settings like `MCP_SERVER_REQUEST_TIMEOUT`. This limitation forces users to work with default timeout values, which may not be suitable for all use cases, especially when dealing with slower MCP servers or long-running operations. This PR adds support to accept the mcp server request's configs to be set from CLI in form of flags and values.

The below configs have been added
| Setting                                 | Description                                                                                                  | Default |
| --------------------------------------- | ------------------------------------------------------------------------------------------------------------ | ------- |
| `MCP_SERVER_REQUEST_TIMEOUT`            | Timeout for requests to the MCP server (ms)                                                                  | 10000   |
| `MCP_REQUEST_TIMEOUT_RESET_ON_PROGRESS` | Reset timeout on progress notifications                                                                      | true    |
| `MCP_REQUEST_MAX_TOTAL_TIMEOUT`         | Maximum total timeout for requests sent to the MCP server (ms) (Use with progress notifications)             | 60000   |

Sample usage for a remote everything-server:
```shell
mcp-inspector-cli --cli http://localhost:3001/sse --method tools/list --request-timeout 1 --reset-timeout-on-progress true --max-total-timeout 5000
```

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
- I have tested the `MCP_SERVER_REQUEST_TIMEOUT` with listing tools as below. Successfully passes or fails according to the timeout duration given.
```
mcp-inspector-cli --cli http://localhost:3001/sse --method tools/list --request-timeout 1 --reset-timeout-on-progress true --max-total-timeout 5000
```

- Tested `MCP_REQUEST_MAX_TOTAL_TIMEOUT` with calling a tool (`longRunningOperation`) from `everything-server` as below. But not able to test the failure case for this.
```
mcp-inspector-cli \  --cli http://localhost:3001 \
  --method tools/call \
  --tool-name longRunningOperation \
  --tool-arg duration=10 \
  --tool-arg steps=10 \
  --request-timeout 1000 \
  --reset-timeout-on-progress true \
  --max-total-timeout 1
```

- `MCP_REQUEST_TIMEOUT_RESET_ON_PROGRESS` has been implemented the same way but not able to test the failure scenarios

## Breaking Changes
<!-- Will users need to update their code or configurations? -->
This PR does not introduce any breaking changes

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [X] My code follows the repository's style guidelines
- [X] New and existing tests pass locally
- [X] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->
The implementation of passing the 3 server configurations has been done like how its done for the mcp web mode. Checked the implementation of `connect` (class `Protocol`) and `request` (class `Client`) methods in `typescript-sdk` repo of  mcp as well. In both cases `RequestOptions` are accepted the same way.

 Could you please suggest further ways to test `MCP_REQUEST_MAX_TOTAL_TIMEOUT` and `MCP_REQUEST_TIMEOUT_RESET_ON_PROGRESS` ?

Kavyapriya-1804 committed:
feat - incorporated mcp server request configs for cli mode

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
synced the PR @olaservo @cliffhall @pulkitsharma07

olaservo commented:
Hi @Kavyapriya-1804 I think one way we can run tests on these changes is to add a script similar to this one: `cli\scripts\cli-tests.js`

If you look at the package.json for the cli folder, the `test` script is running that script.  It runs specific testing scenarios through the CLI using the Everything server as a test server, which is one we use a lot for validation.

I created an example [here](https://gist.github.com/olaservo/1f6fd139a9ae0499d98b0afccb82f060).  The only issue I had with this example is that I can't get the `Test 5: Reset timeout on progress disabled - should fail with timeout` scenario to fail as expected, but that could be an issue with the test script and not the implementation.  Hopefully it gives you an idea of how to test this though.

I was running the above example script by adding it to the package.json in the cli folder, eg `"test:timeouts": "node scripts/timeout-tests.js",`

KavyapriyaJG committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Hi @olaservo,
Thanks for the test cases !

I was able to run them successfully. But wasn't able to get 2/ 9 test cases right (`Test 4: Reset timeout on progress enabled - should complete successfully` and `Max total timeout exceeded (should fail)`). 
I guess this is because we are not able to capture the progress notifications from the CLI. When I looked into the CLI implementation of `tools/call` too, I wasn't able to find any code related to capturing `notifications` of a progress.

Kindly suggest further..

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Hi @cliffhall , @olaservo, @pulkitsharma07 
I don't see any implementation in cli that captures and logs `notifications` when using CLI mode in terminal.
Though I could see **server notifications** being listened and shown in **UI in inspector web** as below.

<img width="1728" alt="Screenshot 2025-05-12 at 23 12 11" src="https://github.com/user-attachments/assets/eb513f1f-04d4-4a6e-922a-a1fdc9f6b47f" />



Implementing this feature could be helpful to receive updates likes `messages` and `progress` on **long running tool calls**.

Would you like to have them implemented as part of this PR ? 
What are your thoughts on this ?

Kavyapriya-1804 commented:
Sure @cliffhall. 
Will analyze and extend the proposal for including notification in the output stream

cc: @olaservo

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
Merge remote-tracking branch 'upstream/main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Hi @olaservo, yeah sure. Would work on the notifications part in a separate issue. 
For now, the code is good with passing  7/9 test cases you gave. Quoting the below detail from our previous conversation ..

> I was able to run them successfully. But wasn't able to get 2/ 9 test cases right (`Test 4: Reset timeout on progress enabled - should complete successfully` and `Max total timeout exceeded (should fail)`).
> I guess this is because we are not able to capture the progress notifications from the CLI. When I looked into the CLI implementation of `tools/call` too, I wasn't able to find any code related to capturing `notifications` of a progress.

Please merge if it looks good to you. Thanks !

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

cliffhall committed:
Merge branch 'main' into feat-cli-server-request-configs

cliffhall commented:
@Kavyapriya-1804 With the command you included, I get the following:
<img width="1548" alt="Screenshot 2025-05-30 at 6 51 33 PM" src="https://github.com/user-attachments/assets/421a42b2-59ce-40e4-b86d-8b815f80b820" />

Can you give a very specific  set of instructions for testing against [server-everything ](https://github.com/modelcontextprotocol/servers/tree/main/src/everything). I.e., setup and test cases for stdio, streamableHttp, and sse, with the various settings and whether failure or success is expected? I don't want to have to guess about this. It would be good if you included screenshots of the results of each test on your system, so that we can compare our results to yours in each case.

Kavyapriya-1804 committed:
Merge remote-tracking branch 'upstream/main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Sure, will add @cliffhall

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
feat - Added test scripts for the CLI server configs

Kavyapriya-1804 commented:
Hi @cliffhall, @olaservo 
Apologies. Caught up in a busy work. I have committed the test script file that @olaservo suggested. 
There were 2 test cases (mentioned below) related to resetTimeout and maxTotalTimeout which I was not able to get up and running because of the missing progress update notification feature in CLI mode. 

1. Test 4: Reset timeout on progress enabled - should complete successfully
2. Test 6: Max total timeout exceeded (should fail)

Could you please help this proceed forward from here ?

cliffhall reviewed (commented):
@Kavyapriya-1804 Why not incorporate these tests into the general cli tests so there's only one suite to run?

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
feat - Moved timeout tests to cli tests

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

cliffhall commented on cli/scripts/cli-tests.js:
```diff
@@ -96,15 +96,93 @@ try {
   );
 
   console.log(`\n${colors.YELLOW}Testing: ${testName}${colors.NC}`);
   TOTAL_TESTS++;
+  console.log("basic test kavya " + TOTAL_TESTS);
```
Remove unnecessary logging. And var declaration at line 27, since it's only used in these log outputs

Kavyapriya-1804 replied:
apologies. removed this test logging.

cliffhall replied:
You removed the logging but left the `TOTAL_TESTS` var at line 27 and it's incrementing at what was lines 184 and 295. Since it was only used for the logging it should be removed as well.

cliffhall commented on cli/scripts/cli-tests.js:
```diff
@@ -177,14 +285,15 @@ async function runBasicTest(testName, ...args) {
   );
 
   console.log(`\n${colors.YELLOW}Testing error case: ${testName}${colors.NC}`);
   TOTAL_TESTS++;
+  console.log("error test kavya " + TOTAL_TESTS);
```
Remove unnecessary logging. And var declaration at line 27, since it's only used in these log outputs

Kavyapriya-1804 replied:
removed this test logging as well

Kavyapriya-1804 replied:
Hi @cliffhall,
This `TOTAL_TESTS` variable exists at 4 places [ initialization, incrementing (for positive, negative cases) and display finally]. And is exactly like how it is in main branch today.

cliffhall reviewed (dismissed):
Remove extra comments

Kavyapriya-1804 committed:
fix - removed test logs

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Hi @cliffhall ,
That might be a node version issue or due to missing dependencies.
I just npm installed at root and cli levels.
It seems to run all the test cases fine.

<img width="646" alt="Screenshot 2025-06-21 at 17 57 30" src="https://github.com/user-attachments/assets/91867f45-5f4e-4acc-8a95-8e3f5bf47a0a" />

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Hi @olaservo,
I see the conflicts from the 2 files. 
Yeah, I really appreciate if you could resolve those conflicts. Thanks !

olaservo committed:
Merge from main

olaservo committed:
Merge branch 'main' into feat-cli-server-request-configs

olaservo committed:
Fix formatting

olaservo committed:
Update cli-tests.js

olaservo committed:
Merge branch 'feat-cli-server-request-configs' of https://github.com/Kavyapriya-1804/inspector into feat-cli-server-request-configs

cliffhall commented:
@Kavyapriya-1804 all the tests are running and I think the changes look good, but one outstanding thing is:

>  You removed the logging but left the TOTAL_TESTS var at line 27 and it's incrementing at what was lines 184 and 295. Since it was only used for the logging it should be removed as well.

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 commented:
Updated this branch with latest from main.
Guess its good to merge now @cliffhall, @olaservo

cliffhall committed:
Merge branch 'main' into feat-cli-server-request-configs

cliffhall reviewed (approved):
LGTM! 👍

cliffhall reviewed (dismissed):
We could use some documentation of how to set these request configs in the README.md.

Kavyapriya-1804 committed:
Merge branch 'main' into feat-cli-server-request-configs

Kavyapriya-1804 committed:
feat: updated documentation for CLI req timeout configs

olaservo committed:
Merge branch 'main' into feat-cli-server-request-configs

cliffhall reviewed (changes requested):
Documentation tweak and questions about missing tests.

Kavyapriya-1804 commented on cli/scripts/cli-tests.js:
```diff
@@ -742,6 +922,138 @@ async function runTests() {
+    "100",
+  );
+
+  // console.log(
+  //   `\n${colors.YELLOW}=== Running Progress-Related Timeout Tests ===${colors.NC}`,
+  // );
+
+  // // Test 32: Reset timeout on progress disabled - should fail with timeout
+  // await runErrorTest(
+  //   "reset_timeout_disabled",
+  //   "should_timeout_quickly",
+  //   TEST_CMD,
+  //   ...TEST_ARGS,
+  //   "--cli",
+  //   "--method",
+  //   "tools/call",
+  //   "--tool-name",
+  //   "longRunningOperation",
+  //   "--tool-arg",
+  //   "duration=15", // Same configuration as above
+  //   "steps=5",
+  //   "--request-timeout",
+  //   "2000",
+  //   "--reset-timeout-on-progress",
+  //   "false", // Only difference is here
+  //   "--max-total-timeout",
+  //   "30000",
+  // );
```
I was not able to get the test implementations right for these 2 params which I have been informing right from the initial days of this PR. It has been 2 months and a lot of code changes, so I am not sure of the current progress for this change.

cliffhall replied:
It would be good to see a test that shows  `--request-timeout` and `--reset-timeout-on-progress true` combined with `--max-total-timeout` actually works. It's the key to managing long running operations.

olaservo committed:
Merge latest changes from main into pr-373

olaservo committed:
Fix cli-tests: add missing null expectedBehavior parameter to test calls

olaservo committed:
Address cliffhall's review: update docs and add combined timeout test

- Update README example to show --max-total-timeout used with --request-timeout and --reset-timeout-on-progress
- Add Test 32: combined_timeout_options that demonstrates all three flags working together
- Update test numbers (34-36) to account for new test
- Add explanatory comment about progress notification limitation in CLI mode
- Fix Windows npx spawn issue by adding shell: true
- Adjust test timeouts to work within 10-second test timeout limit
- Document that progress-based timeout resets don't work in CLI mode yet

olaservo committed:
Merge branch 'main' into feat-cli-server-request-configs

cliffhall committed:
Merge branch 'main' into feat-cli-server-request-configs
